tools/cldr/cldr-to-icu/src/main/java/org/unicode/icu/tool/cldrtoicu/AlternateLocaleData.java

```diff
@@ -0,0 +1,126 @@
+// © 2019 and later: Unicode, Inc. and others.
+// License & terms of use: http://www.unicode.org/copyright.html
+package org.unicode.icu.tool.cldrtoicu;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static org.unicode.cldr.api.CldrDataType.LDML;
+
+import java.util.Map;
+import java.util.Set;
+
+import org.unicode.cldr.api.CldrData;
+import org.unicode.cldr.api.CldrDataSupplier;
+import org.unicode.cldr.api.CldrDataType;
+import org.unicode.cldr.api.CldrDraftStatus;
+import org.unicode.cldr.api.CldrPath;
+import org.unicode.cldr.api.CldrValue;
+
+import com.google.common.collect.ImmutableMap;
+
+/**
+ * A factory for data suppliers which can filter CLDR values by substituting values from one path
+ * to another. The replaced value must retain the original "target" path but will have the value
+ * and value attributes of the "source". A value will only be replaced if both the source and
+ * target paths have associated values. The replacement retains its original position in the value
+ * ordering.
+ *
+ * <p>This class DOES NOT transform supplemental or BCP-47 data, because the use of "alt" values
+ * is completely different for that data (it would require merging specific attributes together).
+ *
+ * <p>Note that this is not a general purpose transformation of CLDR data, since it is generally
+ * not possible to "move" values between arbitrary paths. Target and source paths must be in the
+ * same "namespace" (i.e. share the same element names) but attributes can differ.
+ *
+ * <p>Note also that the mapping is not recursive, so mapping {@code A -> B} and {@code B -> C}
+ * will NOT cause {@code A} to be mapped to {@code C}.
+ *
+ * <p>Typically this class is expected to be used for selecting alternate values of locale data
+ * based on the {@code "alt"} path attribute (e.g. selecting the short form of a region name).
+ */
+public final class AlternateLocaleData {
```

Comment:
This is a neat abstraction to avoid any knowledge of alt-paths in the core logic. It returns a supplier which modifies CLDR data on the fly based on a given map. This means that all mapping classes just see a "normal" supplier and don't have to care.